File: algorithm-problems/src/main/java/us/supercheng/algorithm/problems/leetcode/convert1darrayinto2darray/Solution.java
package us.supercheng.algorithm.problems.leetcode.convert1darrayinto2darray;

public class Solution {

    public int[][] construct2DArray(int[] original, int m, int n) {
        if (original == null || original.length != m * n)
            return new int[0][0];

        int[][] ret = new int[m][n];
        int idx = 0;

        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                ret[i][j] = original[idx++];

        return ret;
    }

    public int[][] construct2DArrayOPT(int[] original, int m, int n) {
        if (original == null || original.length != m * n)
            return new int[0][0];

        int[][] ret = new int[m][n];

        for (int i = 0; i < m; i++)
            System.arraycopy(original, i * n, ret[i], 0, n);

        return ret;
    }

}

File: algorithm-problems/src/main/java/us/supercheng/algorithm/problems/leetcode/convert1darrayinto2darray/Solution.py
class Solution(object):
    def construct2DArray(self, original, m, n):
        """
        :type original: List[int]
        :type m: int
        :type n: int
        :rtype: List[List[int]]
        """
        if m * n != len(original):
            return []

        idx = 0
        ret = []

        for i in range(m):
            l = []
            for j in range(n):
                l.append(original[idx])
                idx += 1
            ret.append(l)

        return ret

def construct2DArrayOPT(self, original, m, n):
        """
        :type original: List[int]
        :type m: int
        :type n: int
        :rtype: List[List[int]]
        """
        if m * n != len(original):
            return []

        ret = []

        for i in range(m):
            ret.append(original[i * n : i * n + n])

        return ret
